src/utils/progress_utils.py

```
@@ -0,0 +1,207 @@
+        
+    except Exception as e:
+        await reporter.report(0, f"{tool_name} failed: {str(e)}", force=True)
+        raise
+
+
+class MultiStageProgressReporter:
+    """
+    Progress reporter for operations with multiple distinct stages.
+    """
+    
+    def __init__(self, ctx: Context, tool_name: str, stages: list):
```

Ruff (0.12.2)

Missing return type annotation for special method `__init__`

Add return type annotation: `None`

(ANN204)